Value-set based dereferencing: fix simplified handling of *(p + i)

tautschnig opened the pull request:
Value-set based dereferencing must not take an access path through an object that precludes a subsequent index expression from accessing a different part of the object. Such a situation can arise when the value set has a known (constant) offset for the pointer that would identify one particular element in an array (within that object). The code using value-set based dereferencing, however, may be trying to resolve a subexpression of an index expression, where said index expression would lead to a different element that may itself be part of a different array within the same overall object.

Fixes: #8570

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Value-set based dereferencing: fix simplified handling of *(p + i)

Value-set based dereferencing must not take an access path through an
object that precludes a subsequent index expression from accessing a
different part of the object. Such a situation can arise when the value
set has a known (constant) offset for the pointer that would identify
one particular element in an array (within that object). The code using
value-set based dereferencing, however, may be trying to resolve a
subexpression of an index expression, where said index expression would
lead to a different element that may itself be part of a different array
within the same overall object.

Fixes: #8570

remi-delmas-3000 commented:
The commit adds a new argument `recurse_into_arrays` to the `get_subexpression_at_offset` to force fallback to using byte-extracts but I could not really understand what condition triggers the fallback ? Is the idea to try to resolve a lookup expression as a cell/field access if enough info is known about the index and default to a flat/unstructured by-extract otherwise ?

remi-delmas-3000 commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -643,11 +643,12 @@ value_set_dereferencet::valuet value_set_dereferencet::build_reference_to(
-      root_object, o.offset(), dereference_type, ns);
-    if(subexpr.has_value())
-      simplify(subexpr.value(), ns);
+      root_object, o.offset(), dereference_type, false, ns);
```
Can you summarize what conditions hold when reaching this line  (i.e. what forces flipping `recurse_into_arrays` to false ?)

remi-delmas-3000 replied:
From the comment right above it rather seems like if recurse_into_arrays is false then you try to build a field access expression, but when it is true you try to build a byte-extract that can hit the array anywhere ?

tautschnig replied:
Thank you for asking all those questions, they just highlighted that the initial change was a bit dubious. Now much simpler and clearer.

tautschnig committed:
get_subexpression_at_offset: remove infeasible branch

The outer decision already ensures that types match, so replacing types
cannot be necessary.

tautschnig commented:
@kroening Reworked as discussed.
